tests/WeakMapTest.php

```diff
@@ -209,14 +205,88 @@ public function testHousekeeping() : void
         unset($k);
         unset($v);
 
-        for ($i = 0; $i < 99; $i++) {
+        if (\PHP_MAJOR_VERSION < 8) {
+            for ($i = 0; $i < 99; $i++) {
+                self::assertNotNull($r->get());
+                isset($weakMap[$unknownObject]);
+            }
+        }
+
+        self::assertNull($r->get());
+    }
+
+    public function testHousekeepingOnGcRun(?WeakMap $weakMap = null) : void
+    {
+        if ($weakMap === null) {
+            $weakMap = new WeakMap();
+        }
+
+        $k = new stdClass;
+        $v = new stdClass;
+        $r = WeakReference::create($v);
+
+        $weakMap[$k] = $v;
+
+        unset($k);
+        unset($v);
+
+        if (\PHP_MAJOR_VERSION < 8) {
             self::assertNotNull($r->get());
-            isset($weakMap[$unknownObject]);
+            gc_collect_cycles();
         }
 
         self::assertNull($r->get());
     }
 
+    public function testNoInternalCycle() : void
+    {
+        $weakMap = new WeakMap();
+        $rWeakMap = WeakReference::create($weakMap);
+
+        $k = new stdClass;
+        $v = new stdClass;
+        $rK = WeakReference::create($k);
+        $rV = WeakReference::create($v);
+
+        $weakMap[$k] = $v;
+
+        unset($weakMap);
+        unset($k);
+        unset($v);
+
+        self::assertNull($rWeakMap->get());
+        self::assertNull($rK->get());
+        self::assertNull($rV->get());
+    }
+
+    public function testHousekeepingOnGcRunSurvival() : void
+    {
+        $weakMap = new WeakMap();
+
+        $vkPairs = [];
+        for ($i = 100; $i > 0; $i--) {
+            for ($j = 100; $j > 0; $j--) {
+                $k = new stdClass;
+                $v = new stdClass;
+                $weakMap[$k] = $v;
+                $vkPairs[] = [$k, $v];
+            }
+
+            gc_collect_cycles();
+            gc_collect_cycles();
+            gc_collect_cycles();
```

Comment:
What's the point of 3 consecutive calls? Do the 2nd and 3rd actually do something?

Reply:
This tests the object cycle survival. 1 - 2 would be enough, but I prefer to keep 3 as I found some php-src related bugs so I prefer to test stronger case.